rfcs/solidity-snapps.md

```diff
@@ -0,0 +1,324 @@
+place of their contents.
+
+In solidity, 'public' variables can be referenced by other contracts via a
+function. We propose using the same method for snapps; see the function section
+below for details.
+
+#### Off-chain storage and snapp state accessibility
+
+When the variables do not fit within the field elements, the data for the snapp
+will not be directly available on-chain, and must be computed or retrieved from
+some off-chain source. It is important to provide primitives for revealing
+the updated states, otherwise updating a snapp's state may only reveal a hash,
+and the new underlying data may be rendered inaccessible.
+
+To this end, it may be useful to add support for the poseidon hash used by mina
+to IPFS, so that this data can be stored (ephemerally) in IPFS. We will also
+discuss a proposal to expose data for state transitions as 'events' associated
+with the parties; see the events section below for details.
+
+### Functions
+
+[Functions](https://docs.soliditylang.org/en/v0.8.5/structure-of-a-contract.html#functions)
+are the primary interface of solidity contracts; in order to interact
+with a smart contract, you submit a transaction that calls one of the functions
+the contract exposes. These may call other functions from the same contract or
+from other contracts.
+
+We propose simulating functions with snark proofs, where each function call
+corresponds to a single snark proof. Our current snark model uses a 'wrapping'
+primitive, which allows a single 'verification key' to verify wrapped proofs
+witnessing one (or more) of several different 'circuits' (here, function
+declarations). Function calls against different snapps require separate
+'parties', although multiple calls to functions in the same snapp may be merged
+into a single proof and issued as a single 'party' (either by proof composition
+or inlining, depending on the use case).
+
+#### Arguments and returned values
+
+In order to simulate calling functions with arguments, and returning values
+from functions, snapp parties must be able to expose some 'witness' to these
+values. The format is determined by the circuit statement, but usually this
+will be `hash(arguments, returned_values)`.
+
+*This is currently not supported by the snapp transaction model RFC.*
+
+**Proposal:** add an additional field element (`aux_data`) that is passed as
+part of the input to the snapp proof, which may be used to bind the function's
+input and returned values.
+
+#### Function calls between snapps
+
+In order for a snapp to verify that function calls are executed, snapp proofs
+must be able to interrogate the other parties included in a transaction. The
+current RFC doesn't identify what the proof inputs should be, but describes a
+stack of parties (`parties`) and the current state of the stack when a
+transaction is reached in the stack (`remaining_parties`).
+
+**Proposal:** pass `parties`, the stack of parties, as part of the snapp input.
```

Comment:
In the implementation, the full input consists of

- the full party stack (hash)
- the party stack at the point of the current party (hash)